### PR TITLE
GitHub Action: Fix authorized-changes-detection

### LINKED ISSUE #1526: GitHub Action: Fix `authorized-changes-detection`
**Is your feature request related to a problem? Please describe.**

1. We have a GitHub action file `authorized-changes-detection.yml` that, during pull requests, detects when sensitive files are modified and exits with an error.
    1. https://github.com/PalisadoesFoundation/talawa-admin/blob/develop/.github/workflows/authorized-changes-detection.yml
1. Unfortunately, as seen in the example PR  below, it does not mention the file(s) that is modifed. This would be very helpful
   1. https://github.com/PalisadoesFoundation/talawa-admin/pull/1488

**Describe the solution you'd like**

1. Migrate the functionality of the `authorized-changes-detection.yml` file to:
    1. https://github.com/PalisadoesFoundation/talawa-admin/blob/develop/.github/workflows/pull-requests.yml
1. The action must:
    1.  Identify when one or more configured directories and files are modified during a PR
    2. List the file(s) in this list that were modified
    3. Exit with failure code.
1. The configuration of the list of sensitive files and directories must be done only once in the file. 
1. Do not use the `paths:` statement. It was suitable for a standalone file, but not for `pull-requests.yml` where other jobs run
1. The `authorized-changes-detection.yml` file must be deleted.
1. The new action must not depend on any other job in the YAML file.

**Describe alternatives you've considered**

We think that this GitHub action could be used
- https://github.com/tj-actions/changed-files

**Approach to be followed (optional)**

- See above

**Additional context**

- N/A

**Potential internship candidates**

Please read this if you are planning to apply for a Palisadoes Foundation internship
- https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by duplixx**
**What kind of change does this PR introduce?**

**Issue Number:**

Fixes #1526 

**Did you add tests for your changes?**

**Snapshots/Videos:**

image

**If relevant, did you update the documentation?**

**Summary**

**Does this PR introduce a breaking change?**

**Other information**

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

### CODE REVIEW

File: .github/workflows/pull-requests.yml [hunk 1]
```diff
@@ -13,6 +13,24 @@ name: PR Workflow
 
 on:
   pull_request:
+    paths:
+      - '.github/**'
+      - 'env.example'
+      - '.husky/**'
+      - 'scripts/**'
+      - 'package.json'
+      - 'tsconfig.json'
```
Comment (palisadoes) [inline .github/workflows/pull-requests.yml:22]: Remove all these paths. This means all the Github actions in this file will only be run when these files are modified. Please research the usage. We stated that we wanted the files defined in only one location

File: .github/workflows/count_changed_files.py [hunk 2]
```diff
@@ -1,18 +1,19 @@
 #!/usr/bin/env python3
 # -*- coding: UTF-8 -*-
-"""Script to limit number of file changes in single PR.
+"""Script to limit the number of file changes in a single PR.
```
Comment (palisadoes) [inline .github/workflows/count_changed_files.py:3]: Rename the script `check_files.py`

File: .github/workflows/sensitive_files.txt [hunk 3]
```diff
@@ -0,0 +1,15 @@
+.github/
+env.example
+.husky/
+scripts/
+package.json
+tsconfig.json
+.gitignore
```
Comment (palisadoes) [inline .github/workflows/sensitive_files.txt:7]: The list here must exactly match what was in the original `paths:` key. You have changed the functionality

**Review by palisadoes — CHANGES_REQUESTED** [review]
Test the following with this PR to make sure this and all other tests run:

1. A commit with changes to more than 20 files not in the list
1. A commit with a test file added to the `.github` directory

**Comment by palisadoes** [comment]
You haven't done the tests I requested using this PR

**Comment by duplixx** [comment]
> You haven't done the tests I requested using this PR

@palisadoes I apologize for any confusion. I am currently running tests, and I have noticed an issue related to the changed files. When printing the changed files, it shows an empty array. I believe this might be the reason why the sensitive file function is not running in this condition. Is there any change i should do to make it work


image
image

**Comment by duplixx** [comment]
@palisadoes i changed the fetch-depth to 1 so it can clone the file name but that also deosn't seems to bew working

**Comment by ratankods** [comment]
taken reference from here

https://stackoverflow.com/questions/63358956/how-to-get-file-names-of-files-updated-by-commit-in-github-actions

https://github.com/actions/checkout?tab=readme-ov-file#checkout-head

**Comment by palisadoes** [comment]
1. Please test more thoroughly. The tests are passing when you have edited files in the `.github/` directory.
2. How will you test if files have been deleted in the paths specified? Is that possible with this methodology, or will it need to be modified or some other test used?

**Comment by duplixx** [comment]
Hey @palisadoes, I'm currently looking for a solution to retrieve the names of lists of modified files using GitHub Actions. Once that's sorted, everything should work fine. Regarding the second point, we haven't added any checks for deleted files yet, but as far as I know, the modified check should also cover deleted files.

**Comment by palisadoes** [comment]
1. Please fix the conflicting file
2. `.github/workflows/pull-requests.yml`  has been recently renamed too

**Comment by palisadoes** [comment]
There is still a conflict. Please investigate

**Comment by duplixx** [comment]
> There is still a conflict. Please investigate

@palisadoes 
Yeah, I'll fix it. Currently, I'm having my end-term exams, so I'm not getting enough time to investigate properly. I'll look into it after the 17th.

**Comment by duplixx** [comment]
@palisadoes here's the working demo of unauthorized changes workflow integrared in pull_request.yml


https://github.com/PalisadoesFoundation/talawa-admin/assets/90516956/35d99535-7248-4261-82fd-af310ff39281

File: .github/workflows/check_files.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    Raises:
+        SystemExit: If an error occurs during execution.
+    """
+
+    args = _arg_parser_resolver()
+
+    base_branch = args.base_branch
+    pr_branch = args.pr_branch
+
+    print(f"You are trying to merge on branch: {base_branch}")
```
Comment (palisadoes) [inline .github/workflows/check_files.py:128]: Only print non error messages if an argparse `--verbose`flag is set. Do this for all cases

File: .github/workflows/check_files.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+    print(f"You are making a commit from your branch: {pr_branch}")
+
+    # Read sensitive files from the provided file
+    with open(args.sensitive_files, 'r') as sensitive_file:
+        sensitive_files = [line.strip() for line in sensitive_file]
+
+    # Count changed files and check for unauthorized changes
+    result = _count_changed_files(base_branch, pr_branch)
```
Comment (palisadoes) [inline .github/workflows/check_files.py:136]: Make `result` have a more meaningful name for readability

File: .github/workflows/check_files.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    Args:
+        changed_files (list): List of changed files.
+        sensitive_files (list): List of sensitive files.
+    Returns:
+        list: Unauthorized changes.
+
+    """
+
+    unauthorized_changes = [file for file in changed_files if any(glob.fnmatch.fnmatch(file, sf) for sf in sensitive_files if sf)]
+    return ScriptResult(len(unauthorized_changes), unauthorized_changes)
```
Comment (palisadoes) [inline .github/workflows/check_files.py:70]: It doesn't make sense to return a named tuple if you can infer one attribute from the other. Just return the files found as a list. I recommended this before for another reason but the code has changed

File: .github/workflows/check_files.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    # Count changed files and check for unauthorized changes
+    result = _count_changed_files(base_branch, pr_branch)
+    print(f"Number of changed files: {result.file_count}")
+
+    unauthorized_changes = _check_unauthorized_changes(result.unauthorized_changes, sensitive_files)
+    print(f"Unauthorized changes: {unauthorized_changes}")
+
+    # Check if the count exceeds the allowed limit
+    if result.file_count > args.file_count:
+        print("Error: Too many files (greater than 20) changed in the pull request.")
```
Comment (palisadoes) [inline .github/workflows/check_files.py:144]: 1. `20` should be replaced with the argparse variable
1. List the number of offending files here too as part of the error message

File: .github/workflows/check_files.py [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+        print("Possible issues:")
+        print("- Contributor may be merging into an incorrect branch.")
+        print("- Source branch may be incorrect; please use 'develop' as the source branch.")
```
Comment (palisadoes) [inline .github/workflows/check_files.py:147]: This should only be displayed if the wrong branch is the target. The logic only prints the message of to many files are modified

File: .github/workflows/check_files.py [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+        sys.exit(1)
```
Comment (palisadoes) [inline .github/workflows/check_files.py:148]: Exit with (1) if any of the test conditions are violated. The user must be able to see all the problems encountered. This way they don't have to keep repeating commits fixing one problem at a time

File: .github/workflows/check_files.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+
+    if unauthorized_changes:
+        print("Error: Unauthorized changes detected. Please review the list of modified files:")
+        for file in unauthorized_changes:
+            print(f" - {file}")
```
Comment (palisadoes) [inline .github/workflows/check_files.py:153]: You are printing the named tuple, not the files it contained. I've recommended another approach

File: .github/workflows/check_files.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+        sys.exit(1)
+    else:
+        print("No unauthorized changes detected. Test passed.")
```
Comment (palisadoes) [inline .github/workflows/check_files.py:156]: Only print non error messages if there is a verbose argparse flag, in all cases

**Review by palisadoes — CHANGES_REQUESTED** [review]
See comments

File: .github/workflows/check_files.py (hunk 8 above)
Comment (duplixx) [reply .github/workflows/check_files.py:147]: Should I go with this approach @palisadoes 
image

Comment (palisadoes) [reply .github/workflows/check_files.py:147]: If any error condition is found then:

1. The error text for each must be displayed
2. When the text for all errors has been displayed, then `sys.exit(1)`. In other words there must be only one `sys.exit(1)` or equivalent exit with error statement.

**Comment by palisadoes** [comment]
We recently updated `.github/workflows/authorized-changes-detection.yml`

**Comment by duplixx** [comment]
> We recently updated `.github/workflows/authorized-changes-detection.yml`

it will be deleted afterwards right??

> We recently updated `.github/workflows/authorized-changes-detection.yml`

added mentioned changes

**Comment by palisadoes** [comment]
Yes, ` .github/workflows/authorized-changes-detection.yml` should be deleted

File: .github/workflows/check_files.py [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+            command, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True
+        )
+        output, error = process.communicate()
+    except Exception as e:
+        print(f"Error: {e}")
+        sys.exit(1)
+
+    changed_files = [file.strip() for file in output.split(' ') if file.strip()]
+    file_count = len(changed_files)
+    return ScriptResult(file_count=file_count, unauthorized_changes=changed_files)
```
Comment (palisadoes) [inline .github/workflows/check_files.py:38]: Again, a named tuple isn't needed here. You are only using the file count later.

File: .github/workflows/check_files.py [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+    if base_branch != "develop":
+        errors.append("Error: You are trying to merge into '{base_branch}' branch.\n  - Source branch may be incorrect; please use 'develop' as the source branch.")
+
+    # Read sensitive files from the provided file
+    with open(args.sensitive_files, 'r') as sensitive_file:
+        sensitive_files = [line.strip() for line in sensitive_file]
+
+    # Count changed files and check for unauthorized changes
+    changed_files = _count_changed_files(base_branch, pr_branch)
+    print(f"Number of changed files: {changed_files.file_count}")
```
Comment (palisadoes) [inline .github/workflows/check_files.py:129]: Only print if the verbose flag is set

**Comment by duplixx** [comment]
All changes are done from my side sir @palisadoes

**Review by palisadoes — CHANGES_REQUESTED** [review]
Make sure the script:

1. is python black, pylint, pycodestyle, flake8 and pydocstyle compliant.
2. Has a maximum line width of 80 for better readability. 
    1. https://stackoverflow.com/questions/59183863/in-python-how-to-tweak-black-formatter-if-possible

**Review by palisadoes — DISMISSED** [review]

**Comment by palisadoes** [comment]
Please fix the conflicting file, so that this can be merged

**Comment by duplixx** [comment]
Sir, @palisadoes, I have deleted this file. How should I fix the conflicts? I guess after merging the changes, the conflicts will not show up.

**Comment by palisadoes** [comment]
Try merging with the latest upstream and updating with a commit

**Comment by palisadoes** [comment]
_**NOTE**_ Read very carefully

1. We just merged this PR which upgraded the prettier package. 
    1. https://github.com/PalisadoesFoundation/talawa-admin/pull/1599
1. It reformatted over 150 files. 

This will put your PR at risk of extensive merge conflicts. Do the following **IN THIS ORDER**:

1. upgrade your `prettier` in your local branch to the same version
1. run `prettier` on your local branch
1. update your local branch with the latest upstream
    
This will help to reduce the number of **existing** and **future** merge conflicts for your PR.

**Review by palisadoes — APPROVED** [review]

**Pull request merged**